providers/implementations/ciphers/cipher_aes_cbc_hmac_sha.c

```diff
@@ -0,0 +1,335 @@
+     */
+    p = OSSL_PARAM_locate_const(params, OSSL_CIPHER_PARAM_TLS1_MULTIBLOCK_AAD);
+    if (p != NULL) {
+        p1 = OSSL_PARAM_locate_const(params,
+                OSSL_CIPHER_PARAM_TLS1_MULTIBLOCK_INTERLEAVE);
+        if (p->data_type != OSSL_PARAM_OCTET_STRING
+            || p1 == NULL
+            || !OSSL_PARAM_get_uint(p1, &mb_param.interleave)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER);
+            return 0;
+        }
+        mb_param.inp = p->data;
+        mb_param.len = p->data_size;
+        if (hw->tls1_multiblock_aad(vctx, &mb_param) <= 0)
+            return 0;
+    }
+
+    /*
+     * The inputs to tls1_multiblock_encrypt are:
+     *   mb_param->inp
+     *   mb_param->len
+     *   mb_param->interleave
+     *   mb_param->out
+     * The outputs of tls1_multiblock_encrypt are:
+     *   ctx->multiblock_encrypt_len
+     */
+    p = OSSL_PARAM_locate_const(params, OSSL_CIPHER_PARAM_TLS1_MULTIBLOCK_ENC);
+    if (p != NULL) {
+        p1 = OSSL_PARAM_locate_const(params,
+                OSSL_CIPHER_PARAM_TLS1_MULTIBLOCK_INTERLEAVE);
+        pin = OSSL_PARAM_locate_const(params,
+                   OSSL_CIPHER_PARAM_TLS1_MULTIBLOCK_ENC_IN);
+        if (p->data_type != OSSL_PARAM_OCTET_STRING
+            || pin == NULL
+            || pin->data_type != OSSL_PARAM_OCTET_STRING
+            || p1 == NULL
+            || !OSSL_PARAM_get_uint(p1, &mb_param.interleave)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER);
+            return 0;
+        }
+        mb_param.out = p->data;
+        mb_param.inp = pin->data;
+        mb_param.len = pin->data_size;
+        if (hw->tls1_multiblock_encrypt(vctx, &mb_param) <= 0)
+            return 0;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CIPHER_PARAM_AEAD_TLS1_AAD);
+    if (p != NULL) {
+        if (p->data_type != OSSL_PARAM_OCTET_STRING) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER);
+            return 0;
+        }
+        if (hw->set_tls1_aad(ctx, p->data, p->data_size) <= 0)
+            return 0;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CIPHER_PARAM_KEYLEN);
+    if (p != NULL) {
+        size_t keylen;
+
+        if (!OSSL_PARAM_get_size_t(p, &keylen)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER);
+            return 0;
+        }
+        if (ctx->base.keylen != keylen) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH);
+            return 0;
+        }
+    }
+    return ret;
+}
+
+static int aes_get_ctx_params(void *vctx, OSSL_PARAM params[])
+{
+    PROV_AES_HMAC_SHA_CTX *ctx = (PROV_AES_HMAC_SHA_CTX *)vctx;
+    PROV_CIPHER_HW_AES_HMAC_SHA *hw =
+       (PROV_CIPHER_HW_AES_HMAC_SHA *)ctx->hw;
+    OSSL_PARAM *p;
+
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_TLS1_MULTIBLOCK_MAX_BUFSIZE);
+    if (p != NULL) {
+        size_t len;
+
+        if (!OSSL_PARAM_get_size_t(p, &len)) {
```

Comment:
Why are we "getting" a len value here? This is a "get" function so we wouldn't normally expect input from the parameters??

Reply:
Looks like a copy-n-paste that needs some cleanup...  eh, @slontis?

Reply:
If you look down to the next line it does a set...
This is really a ctrl function that returns something - so its a get and a set..
Do you want it to be a set - either way its confusing :)

Reply:
Oh, you made it an input *and* output parameter in one go! ...  and yeah, I see what you're doing there, you're *getting* a value given an input value.

I can accept that, although I wish I hadn't seen it...

Reply:
I suggest adding a comment, 'cause anyone looking at this will wonder, just like @mattcaswell did

Reply:
Hmmmmmm.....I'm not really sure that is a good idea :-(

Reply:
Me neither, but I can't see a way around it for the moment

Reply:
The input is the max_send_fragment. And the output is the size of that fragment after overheads have been applied. So this seems to be a set operation (to set the max_send_fragment) followed by a get. The provider should implement it that way as two separate operations. In the EVP layer where we implement the ctrl we just make two calls - first a set and then then get.

Reply:
ok - done (it means another value needs to be cached - but at least its consistent)..